qemu/tests/throttle_multi_guests_parameter_test.py

```diff
@@ -0,0 +1,98 @@
+import os
+import logging
+import shutil
+
+from virttest import error_context
+from virttest import qemu_storage
+from virttest import data_dir
+from virttest import env_process
+
+from threading import Thread
+
+from provider.storage_benchmark import generate_instance
+from provider.throttle_utils import ThrottleGroupManager, ThrottleTester, \
+    ThrottleGroupsTester
+
+
+# This decorator makes the test function aware of context strings
+@error_context.context_aware
+def run(test, params, env):
+    """
+        Test throttle relevant properties feature.
+
+        1) Boot up guest with throttle groups.
+        There are two throttle groups and each have two disk
+        2) Build fio operation options and expected result
+         according to throttle properties.
+        3) Execute one disk or all disks testing on groups parallel.
+    """
+
+    def copy_base_vm_image():
+        """Copy the base vm image for VMs."""
+        src_img = qemu_storage.QemuImg(
+            params, data_dir.get_data_dir(), params['images'])
+        src_filename = src_img.image_filename
+        src_format = src_img.image_format
+        dst_dir = os.path.dirname(src_filename)
+        for vm_name in vms_list:
+            dst_filename = os.path.join(
+                dst_dir, '%s.%s' % (vm_name, src_format))
+            if not os.path.exists(dst_filename):
+                logging.info('Copying %s to %s.', src_filename, dst_filename)
+                shutil.copy(src_filename, dst_filename)
+
+    def wait_for_login_all_vms():
+        """Wait all VMs to login."""
+        return [vm.wait_for_login() for vm in vms]
+
+    @error_context.context_aware
+    def fio_on_vm(vm_t, session_t):
+        error_context.context("Deploy fio", logging.info)
+        fio = generate_instance(params, vm_t, 'fio')
+        logging.info("fio: %s" % fio)
+        tgm = ThrottleGroupManager(vm_t)
+        logging.info("tgm: %s" % tgm)
+        groups = params["throttle_groups"].split()
+        testers = []
+        for group in groups:
+            tgm.get_throttle_group_props(group)
+            images = params["throttle_group_member_%s" % group].split()
+            tester = ThrottleTester(test, params, vm_t, session_t,
+                                    group, images)
+            error_context.context("Build test stuff for %s:%s"
+                                  % (group, images), logging.info)
+            tester.build_default_option()
+            tester.build_images_fio_option()
+            tester.set_fio(fio)
+            testers.append(tester)
+        error_context.context("Start groups testing:%s" % groups, logging.info)
+        groups_tester = ThrottleGroupsTester(testers)
+        groups_tester.start()
+
+    def fio_on_vms():
```

Comment:
what about using utils_misc.parallel() here instead of Thread, parallel has already called join in the end.

Reply:
The more we use VT to do the tasks, the simpler and shorter our code could be, just fyi.

Reply:
zixi, one more thing, python's thread can only be used for tasks with heavy IOs, in your case, please make sure that fio inside vms can be run concurrently(like task1->task2->task1...), if not, please use multiple processing instead

Reply:
Thanks for your comment, change to utils_misc.parallel().
Also confirmed fio can run concurrently insides vms. As a fio cmd will run 60s, so it is ok to use multi-thread to simulate two vms doing fio at the same time. We just need a moment that have two vms doing fio concurrently and works well.